tests/e2e/corrupt_test.go

```diff
@@ -98,6 +98,79 @@ func corruptTest(cx ctlCtx) {
 	e2e.WaitReadyExpectProc(context.TODO(), proc, []string{fmt.Sprintf("etcdmain: %016x found data inconsistency with peers", id0)})
 }
 
+func TestInPlaceRecovery(t *testing.T) {
+	basePort := 20000
+	e2e.BeforeTest(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Initialize the cluster.
+	epcOld, err := e2e.NewEtcdProcessCluster(ctx, t,
+		e2e.WithInitialClusterToken("old"),
+		e2e.WithKeepDataDir(false),
+		e2e.WithCorruptCheckTime(time.Second),
+		e2e.WithBasePort(basePort),
+	)
+	if err != nil {
+		t.Fatalf("could not start etcd process cluster (%v)", err)
+	}
+
+	// Put some data into the old cluster, so that after recovering from a blank db, the hash diverges.
+	oldCc, err := e2e.NewEtcdctl(epcOld.Cfg.Client, epcOld.EndpointsGRPC())
+	assert.NoError(t, err)
+	for i := 0; i < 10; i++ {
+		err := oldCc.Put(ctx, testutil.PickKey(int64(i)), fmt.Sprint(i), config.PutOptions{})
+		assert.NoError(t, err, "error on put")
+	}
+
+	// Create a new cluster config, but with the same port numbers. In this way the new servers can stay in
+	// contact with the old ones.
+	epcNewConfig := e2e.NewConfig(
+		e2e.WithInitialClusterToken("new"),
+		e2e.WithKeepDataDir(false),
+		e2e.WithCorruptCheckTime(time.Second),
+		e2e.WithBasePort(basePort),
+		e2e.WithInitialCorruptCheck(true),
+	)
+	epcNew, err := e2e.InitEtcdProcessCluster(t, epcNewConfig)
+	if err != nil {
+		t.Fatalf("could not init etcd process cluster (%v)", err)
+	}
+	newCc, err := e2e.NewEtcdctl(epcNew.Cfg.Client, epcNew.EndpointsGRPC())
+	assert.NoError(t, err)
+
+	// Rolling recovery of the servers.
+	for i, newProc := range epcNew.Procs {
+		oldProc := epcOld.Procs[i]
+		err = oldProc.Close()
+		if err != nil {
+			t.Fatalf("could not stop etcd process (%v)", err)
+		}
+		err = newProc.Start(ctx)
```

Comment:
Part of the start time is waiting for the member to be ready. This causes the test to wait indefinitely and not really test the inplace restore. Please make sure that the test is able to do an inplace restore with corruption checking disabled and leads to corrupted state. And only then create a separate test that shows that you can prevent corruption by  enabling the checks.

Reply:
I am a little bit confused.

>Part of the start time is waiting for the member to be ready. If I run the test without your fixes do corruption detection, will wait indefinitely for first member to join the cluster. 

Could you please double check or provide some log? I tried to run the test without my fix, and there was no indefinite wait. It reported corruptions as expected, and seemed the new cluster was on by log.

The new cluster (pid43115-43118)
```
{"Time":"2023-05-25T18:48:30.482422-05:00","Action":"output","Test":"TestInPlaceRecovery","Output":"/Users/caojiaming/go/src/etcd/bin/etcd (TestInPlaceRecovery-test-0) (43115): {\"level\":\"info\",\"ts\":\"2023-05-25T18:48:30.482351-0500\",\"logger\":\"raft\",\"caller\":\"v3@v3.0.0-20221201111702-eaa6808e1f7a/raft.go:797\",\"msg\":\"ca50e9357181d758 became leader at term 2\"}\r\n"}
...
{"Time":"2023-05-25T18:48:30.478547-05:00","Action":"output","Test":"TestInPlaceRecovery","Output":"/Users/caojiaming/go/src/etcd/bin/etcd (TestInPlaceRecovery-test-1) (43117): {\"level\":\"info\",\"ts\":\"2023-05-25T18:48:30.478433-0500\",\"logger\":\"raft\",\"caller\":\"v3@v3.0.0-20221201111702-eaa6808e1f7a/raft.go:728\",\"msg\":\"5ac8aa22f1eb4c8f became follower at term 2\"}\r\n"}
...
{"Time":"2023-05-25T18:48:37.571722-05:00","Action":"output","Test":"TestInPlaceRecovery","Output":"/Users/caojiaming/go/src/etcd/bin/etcd (TestInPlaceRecovery-test-2) (43118): {\"level\":\"info\",\"ts\":\"2023-05-25T18:48:37.571695-0500\",\"logger\":\"raft\",\"caller\":\"v3@v3.0.0-20221201111702-eaa6808e1f7a/raft.go:728\",\"msg\":\"72e86eb5e356a737 became follower at term 2\"}\r\n"}
```

and the old cluster is down (pid 43102-43104):
```
{"Time":"2023-05-25T18:48:30.420149-05:00","Action":"output","Test":"TestInPlaceRecovery","Output":"/Users/caojiaming/go/src/etcd/bin/etcd (TestInPlaceRecovery-test-2) (43104): {\"level\":\"warn\",\"ts\":\"2023-05-25T18:48:30.420045-0500\",\"logger\":\"raft\",\"caller\":\"v3@v3.0.0-20221201111702-eaa6808e1f7a/raft.go:1039\",\"msg\":\"665e70146a5ab3b6 stepped down to follower since quorum is not active\"}\r\n"}
```

> Please make sure that the test is able to do an inplace restore with corruption checking disabled and leads to corrupted state. 

Do you mean with the cluster-id check disabled? With corruption checking disabled we cannot even know there's a corruption. If so, should I introduce a command-line arg or environment variable or something that turns on/off the check?

Thank you for you patience in advance.

Reply:
I think I cheated a little bit and used some older etcd binary. Possibly this works differently on current main branch. Give me some time to confirm.

Reply:
@serathius Any updates? Can this be merged? Thanks！

Reply:
LGTM